text/0037-anr-rates.md

```diff
@@ -0,0 +1,107 @@
+
+```bash
+(unique users who experienced error events with tag mechanism: ANR)
+___________________________________________________________________ x 100 
+                    (total unique users)
+```
+
+There are a couple problems with this:
+
+1. Total ANRs is affected by client side sampling and dropped events if org/project is close to error quota
+2. The most accurate count of total unique users will probably come from the sessions dataset
+3. Getting this information will require clickhouse queries to two different datasets and post-processing to calculate the ANR rate
+
+Issues outlined in 1 & 2 will result in us showing *wrong* ANR rates and 3 will limit the capabilities of ANR rate - can’t allow sort and search, can’t add it to dashboards or alerts and is not future-proof in case we want to use ANR rates in issue detection.
+
+# Supporting Data
+
+
+# Options Considered
+
+## Introduce a new tag in the sessions dataset (Option 1)
+
+Introduce a new optional tag `freeze_mechanism` in the sessions dataset and track ANRs in sessions in addition to sending the ANR error events. 
+
+SDK sends a session update (`freeze_mechanism:ANR`) when it hits an ANR in addition to creating an ANR error event.
+
+We will then calculate ANR rate with a query similar to counting abnormal or errored users:
+
+```sql
+SELECT 
+    (uniqCombined64MergeIf((value AS _snuba_value), equals((arrayElement(tags.value, indexOf(tags.key, 'freeze_mechanism')) AS `_snuba_tags[freeze_mechanism]`), 'ANR') AND in((metric_id AS _snuba_metric_id), [user])) AS `_snuba_session.anr_user`), 
+    (uniqCombined64MergeIf((value AS _snuba_value), equals((metric_id AS _snuba_metric_id), 'user')) AS `_snuba_count_unique(sentry.sessions.user)`), 
+    (divide(`_snuba_session.anr_user`,`_snuba_count_unique(sentry.sessions.user)`) AS `_snuba.anr_rate`)
+   FROM metrics_sets_v2_local
+  WHERE equals(granularity, ...)
+    AND equals((org_id AS _snuba_org_id), ...)
+    AND in((project_id AS _snuba_project_id), [...])
+    AND greaterOrEquals((timestamp AS _snuba_timestamp), toDateTime('...', 'Universal'))
+    AND less(_snuba_timestamp, toDateTime('...', 'Universal'))
+    AND in(_snuba_metric_id, [user])
+  LIMIT 2
+```
+
+1. This is a more generic solution that’s extensible to "freeze mechanisms" (ex. app hangs on iOS) from other operating systems
+2. This can also be extended to track the four different types of ANRs outlined [here](https://developer.android.com/topic/performance/vitals/anr). Differentiating ANRs due to `Input dispatching timed out` separately will allow us to calculate user-perceived ANR rate
+3. Stand-alone addition that won’t require any migration, less risky, doesn’t affect existing calculations
+
+## Drawbacks
+
+1. Increases the cardinality of the sessions data, however this tag would only be set by Android SDKs for now, so would only affect Android sessions
```

Comment:
Increasing cardinality is not an issue per se. It increases cost but it is manageable.